airsonic-main/src/main/java/org/airsonic/player/dao/LegacyHsqlDaoHelper.java

```diff
@@ -31,29 +51,226 @@ public void checkpoint() {
+            return true;
+        } else if (currentVersion.startsWith(driverVersion)) {
+            // If we're already on the same version as the driver, nothing should happen.
+            LOG.info("HSQLDB database upgrade unneeded, already on version {}.", driverVersion);
+            return false;
+        } else if (currentVersion.startsWith("2.")) {
+            // If the database version is 2.x but older than the driver, the upgrade should be relatively painless.
+            LOG.info("HSQLDB database will be silently upgraded from version {} to {}.", currentVersion, driverVersion);
+            return false;
+        } else {
+            // If this happens we're on a completely untested version and we don't know what will happen.
+            LOG.warn("HSQLDB database upgrade from version {} to version {} is untested. Trying to connect anyway, this may upgrade the database.", currentVersion, driverVersion);
+            return false;
+        }
+    }
+
+    /**
+     * Perform a backup of the HSQLDB database, to a timestamped directory.
+     * @return the path to the backup directory
+     */
+    private Path performHsqldbDatabaseBackup() throws IOException {
+
+        String timestamp = new SimpleDateFormat("yyyyMMddHHmmss").format(new Date());
+        Path source = Paths.get(settingsService.getDefaultJDBCPath()).getParent();
+        Path destination = Paths.get(String.format("%s.backup.%s", settingsService.getDefaultJDBCPath(), timestamp));
+
+        LOG.info("Backing up HSQLDB database to {}...", destination.toString());
+        try (Stream<Path> stream = Files.walk(source)) {
+
+            List<Path> sources = stream.collect(toList());
+            List<Path> destinations = sources.stream()
+                    .map(source::relativize)
+                    .map(destination::resolve)
+                    .collect(toList());
+
+            for (int i = 0; i < sources.size(); i++) {
+                LOG.info("Copying {} to {}", sources.get(i), destinations.get(i));
+                Path currentSource = sources.get(i);
+                Path currentDestination = destinations.get(i);
+                if (currentDestination.toFile().isDirectory()) {
+                    currentDestination.toFile().mkdirs();
+                } else {
+                    currentDestination.getParent().toFile().mkdirs();
+                    Files.copy(sources.get(i), destinations.get(i));
+                }
+            }
+        }
+
+        LOG.info("HSQLDB database backup complete.");
+        return destination;
+    }
+
+    /**
+     * Perform an in-place database upgrade from HSQLDB 1.x to 2.x.
+     */
+    private void performHsqldbDatabaseUpgrade() throws SQLException {
+        // This will upgrade HSQLDB from 1.8 to 2.5 on the first connection.
+        try (Connection conn = DriverManager.getConnection(settingsService.getDefaultJDBCUrl())) {
```

Comment:
What i'm most curious about is how this is able to establish a connection without a username/password. That jdbc url is protected by a username at least, right?

I would prefer to be safe and replace all this with a `getJdbcTemplate.execute()`

Reply:
In this case the username/password are embedded in the JDBC URL.

Reply:
I don't see it? Where? 
`settingsService.getDefaultJDBCUrl()` points to https://github.com/airsonic/airsonic/pull/1345/files#diff-0291e6bf8d7243eb6dd454bd00f440dfR295
I don't see a user/password in that url

Reply:
Eh, you're actually right, I haven't revisited this since I wrote the code and I spoke too fast.

Not sure how it works, but the username/password used by Airsonic is probably the default in HSQLDB if you don't specify anything? In this case I agree, it makes sense to specify it explicitly. I'll write something to that effect.

Reply:
just use the JdbcTemplate, it has the user/pass from the source inside it. No need to custom write anything else

`getJdbcTemplate().execute("SHUTDOWN SCRIPT");'

Reply:
Fixed by explicitly setting username/password.